### PR TITLE
Create project as host project (shared_vpc_host)

### PR DESCRIPTION
**Pull request by thiagonache**
Hi @morgante @bharathkkb ,

After a lot of discussions, I'm opening a new PR with a clean branch. I hope all is correct now.

### CODE REVIEW

**Comment by thiagonache** [comment]
Not sure why it's failing running make on my local machine it does pass.

**Comment by morgante** [comment]
The error is here:
```
terraform_validate ./test/fixtures/full 

Error: Unsupported argument

  on .terraform/modules/vpc/main.tf line 58, in resource "google_compute_subnetwork" "subnetwork":
  58:   enable_flow_logs         = lookup(var.subnets[count.index], "subnet_flow_logs", "false")

An argument named "enable_flow_logs" is not expected here.
```

**Comment by thiagonache** [comment]
I don’t think it is related to my changes... do you?

**Comment by thiagonache** [comment]
> I don’t think it is related to my changes... do you?
> […](#)
> On 1 Oct 2020, at 16:56, Morgante Pell ***@***.***> wrote: ﻿ The error is here: terraform_validate ./test/fixtures/full Error: Unsupported argument on .terraform/modules/vpc/main.tf line 58, in resource "google_compute_subnetwork" "subnetwork": 58: enable_flow_logs = lookup(var.subnets[count.index], "subnet_flow_logs", "false") An argument named "enable_flow_logs" is not expected here. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

What I've found is that enable_flow_logs was deprecated, we should use log_config instead. Trying to match the same version that is working for me.

**Comment by thiagonache** [comment]
@morgante for some odd reason the google provider version pulled by my terraform execution end up with a different version than in the CI environment. I've deleted the terraform cache and confirmed that it pulled out the version 3.40 again, so I've created the required version config for the provider in fixtures/versions.tf and it worked.
The last thing I've just done is to revert the commit to pin the module version that was my initial suspicion. I'm confident that it will work this time.

**Comment by thiagonache** [comment]
yay!

File: examples/shared_vpc/main.tf [hunk 1]
```diff
@@ -42,13 +42,15 @@ provider "random" {
...
-  billing_account      = var.billing_account
-  skip_gcloud_download = true
+  source                         = "../../"
+  random_project_id              = true
+  name                           = var.host_project_name
+  org_id                         = var.organization_id
+  folder_id                      = var.folder_id
+  billing_account                = var.billing_account
+  skip_gcloud_download           = true
+  shared_vpc                     = var.host_project_name
```
Comment (morgante) [inline examples/shared_vpc/main.tf:52]: We shouldn't need this on the host project.

File: modules/core_project_factory/variables.tf [hunk 2]
```diff
@@ -182,11 +182,17 @@ variable "disable_dependent_services" {
...
 
-variable "shared_vpc_enabled" {
-  description = "If shared VPC should be used"
+variable "enable_shared_vpc_service_project" {
+  description = "If this project should be attached to a shared VPC. If true, you must set shared_vpc variable"
   type        = bool
 }
 
+variable "enable_shared_vpc_host_project" {
+  description = "If this project is a shared VPC host project. If true, you must set shared_vpc variable. Default is false for compatibility"
```
Comment (morgante) [inline modules/core_project_factory/variables.tf:191]: Actually if it's the host project we should *not* set `shared_vpc` variable.

File: modules/gsuite_enabled/variables.tf [hunk 3]
```diff
@@ -171,8 +171,8 @@ variable "disable_dependent_services" {
   type        = string
 }
 
-variable "shared_vpc_enabled" {
-  description = "If shared VPC should be used"
+variable "enable_shared_vpc_service_project" {
```
Comment (morgante) [inline modules/gsuite_enabled/variables.tf:174]: Please leave the name unchanged on outer modules for backwards-compatibility.

File: modules/shared_vpc/variables.tf [hunk 4]
```diff
@@ -165,8 +165,8 @@ variable "disable_dependent_services" {
   type        = bool
 }
 
-variable "shared_vpc_enabled" {
-  description = "If shared VPC should be used"
+variable "enable_shared_vpc_service_project" {
```
Comment (morgante) [inline modules/shared_vpc/variables.tf:168]: Please leave the name unchanged on outer modules for backwards-compatibility.

File: variables.tf [hunk 5]
```diff
@@ -48,6 +48,12 @@ variable "shared_vpc" {
   default     = ""
 }
 
+variable "enable_shared_vpc_host_project" {
+  description = "If this project is a shared VPC host project. If true, you must set shared_vpc variable. Default is false for compatibility"
```
Comment (morgante) [inline variables.tf:52]:
```suggestion
  description = "If this project is a shared VPC host project. If true, you must set *not* shared_vpc variable. Default is false."
```

**Review by morgante — CHANGES_REQUESTED** [review]
When creating the host project you should *not* need a `shared_vpc` input variable with the name.

**Comment by thiagonache** [comment]
> When creating the host project you should _not_ need a `shared_vpc` input variable with the name.

that's the problem with the logic and mentioned. We agreed on that...
the only way to determine the variable is if shared_vpc is set

File: variables.tf (hunk 5 above)
Comment (thiagonache) [reply variables.tf:52]: @morgante actually, no... shared_vpc is the name of the host vpc, this is a boolean.

File: modules/shared_vpc/variables.tf (hunk 4 above)
Comment (thiagonache) [reply modules/shared_vpc/variables.tf:168]: @morgante you suggested renaming this variable. You meant we should have two different names in different places for the same variable? I'm confused now.

File: modules/gsuite_enabled/variables.tf (hunk 3 above)
Comment (thiagonache) [reply modules/gsuite_enabled/variables.tf:174]: again, you asked to rename this variable. should I do it only in the root module? it's going to be confusing.

File: modules/core_project_factory/variables.tf (hunk 2 above)
Comment (thiagonache) [reply modules/core_project_factory/variables.tf:191]: we must set... remember that shared_vpc == host project name.. that why we opened the issue to rename this variable

File: examples/shared_vpc/main.tf (hunk 1 above)
Comment (thiagonache) [reply examples/shared_vpc/main.tf:52]: again, shared vpc is the host project name... it's better to repeat here than add an if in the code

**Comment by thiagonache** [comment]
@morgante that's the current code
```
shared_vpc                         = var.shared_vpc
shared_vpc_enabled                 = var.shared_vpc != ""
```
is your suggestion is just to change to
```
shared_vpc                                          = var.shared_vpc
enable_shared_vpc_service_project                 = var.shared_vpc != ""
enable_shared_vpc_host_project = var. enable_shared_vpc_host_project
```
This is a lot confusing IMO
Please, confirm.

**Comment by morgante** [comment]
> that's the problem with the logic and mentioned. We agreed on that...
> the only way to determine the variable is if shared_vpc is set

No it's not. There is no reason to set `shared_vpc` on the host project—it already *is* the host project. Just ignore it completely. It's very confusing to pass a name to the `shared_vpc` argument on the host project.

**Comment by thiagonache** [comment]
> > that's the problem with the logic and mentioned. We agreed on that...
> > the only way to determine the variable is if shared_vpc is set
> 
> No it's not. There is no reason to set `shared_vpc` on the host project—it already _is_ the host project. Just ignore it completely. It's very confusing to pass a name to the `shared_vpc` argument on the host project.

yeah... I've removed this comment. Please, see my last comment

**Comment by morgante** [comment]
> @morgante okay. you're right, but I think we should rename shared_vpc_enabled on the next major release as we will do with shared_vpc, do you agree?

Yep.

**Comment by thiagonache** [comment]
Please, see related PR.

**Pull request closed without merging**